### PR TITLE
ML track notes

### PR DESCRIPTION
**Pull request by colinmorris**

### CODE REVIEW

File: learntools/machine_learning/nbs/tut1.ipynb [hunk 1]
```diff
@@ -0,0 +1,78 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "_cell_guid": "69a3a8be-b3cd-45b5-82bc-333e91e173bf",
+    "_uuid": "fa74e8d20864d22e0c5622cdeb4d695bae301948"
+   },
+   "source": [
+    "**[Machine Learning Course Home Page](kaggle.com/learn/machine-learning).**\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut1.ipynb]: You need a `https://` prefix for this, otherwise it gets munged.

File: learntools/machine_learning/nbs/tut1.ipynb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    "![First Decision Trees](http://i.imgur.com/7tsb5b1.png)\n",
+    "\n",
+    "It divides houses into only two categories. You predict the price of a new house by finding out which category it's in, and the prediction is the historical average price from that category.\n",
+    "\n",
+    "This captures the relationship between house size and price. We use data to decide how to break the houses into two groups,  and then again to determine the predicted price in each group.  This step of capturing patterns from data is called **fitting** or **training** the model. The data used to **fit** the model is called the **training data**.  \n",
+    "\n",
+    "The details of how the model is fit (e.g. how to split up the data) is complex enough that we will save it for later. After the model has been fit, you can apply it to new data to **predict** prices of additional homes.\n",
+    "\n",
+    "---\n",
+    "# Example\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut1.ipynb]: The organization here is a bit confusing. We have a heading ("Example"), then the image, then a hr, then another heading ("Improving the Decision Tree"), which makes it seem like this image and the question that goes along with it are supposed to be some self-contained thing. In actuality it seems like the stuff under the hr is a continuation. I'd suggest putting the "Improving the Decision Tree" heading where "Example" currently is, and removing the hr separator.

File: learntools/machine_learning/nbs/tut1.ipynb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    "\n",
+    "You can capture more factors using a tree that has more \"splits.\" These are called \"deeper\" trees. A decision tree that also considers the total size of each house's lot might look like this: \n",
+    "![Depth 2 Tree](http://i.imgur.com/R3ywQsR.png)\n",
+    "\n",
+    "You predict the price of any house by tracing through the decision tree, always picking the path corresponding to that house's characteristics. The predicted price for the house is at the bottom of the tree.  The point at the bottom where we make a prediction is called a **leaf.**   \n",
+    "\n",
+    "The splits and values at the leaves will be determined by the data, so it's time for you to check out the data you will be working with.\n",
+    "\n",
+    "# Continue\n",
+    "You will write code as part of an ongoing data science project for the rest of the tutorials. **[Click here](https://www.kaggle.com/dansbecker/starting-your-ml-project)** to get started."
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut1.ipynb]: I was a little confused by the verbiage here. It seems to suggest that the next thing we're going to do is write some code.

File: learntools/machine_learning/nbs/tut2.ipynb [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "_cell_guid": "b62d4ca0-38c8-4cbf-91f8-d5b95eb293fc",
+    "_uuid": "390958537e6e6f663708b2d2cd4bd42b3a3d4cfd"
+   },
+   "source": [
+    "The most important part of the Pandas library is the DataFrame.  A DataFrame holds the type of data you might think of as a table. This is similar to a sheet in Excel, or a table in a SQL database. The Pandas DataFrame has powerful methods for most things you'll want to do with this type of data.  Let's start by looking at a basic data overview with our example data from Melbourne and the data you'll be working with from Iowa.\n",
+    "\n",
+    "The example will use data at the file path **`../input/melbourne-housing-snapshot/melb_data.csv`**.  Your data will be available in your notebook at **`../input/house-prices-advanced-regression-techniques/train.csv`** (which is already typed into the sample code for you). **Remember that you will be using the Iowa data instead of the Melbourne data when writing your own code.**\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut2.ipynb]: > Remember that you will be using the Iowa data instead of the Melbourne data when writing your own code.

Isn't this the first time that Melbourne and Iowa are mentioned? If so, it seems odd to use the word "remember".

Maybe before talking about Pandas you could have a brief section at the top that frames the problem. Recapping the scenario of wanting to predict housing prices (in order to buy low and sell high), then introducing the two datasets we'll be using, and the fact that the example code will use Melbourne, and the code they write will use Iowa (insert in-universe explanation here).

File: learntools/machine_learning/nbs/tut2.ipynb [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    "\n",
+    "The second value is the **mean**, which is the average.  Under that, **std** is the standard deviation, which measures how numerically spread out the values are.\n",
+    "\n",
+    "To interpret the **min**, **25%**, **50%**, **75%** and **max** values, imagine sorting each column from lowest to highest value.  The first (smallest) value is the min.  If you go a quarter way through the list, you'll find a number that is bigger than 25% of the values and smaller than 75% of the values.  That is the **25%** value (pronounced \"25th percentile\").  The 50th and 75th percentiles are defined analgously, and the **max** is the largest number.\n",
+    "\n",
+    "--- \n",
+    "\n",
+    "\n",
+    "# Your Turn\n",
+    "Get started with your **[first coding exercise](https://www.kaggle.com/kernels/fork/1258954)****\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut2.ipynb]: Extra asterisks

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 6]
```diff
@@ -0,0 +1,184 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This exercise will test your ability to read a data file and understand statistics about the data.\n",
+    "\n",
+    "In later exercises, you will techniques to filter the data, build a machine learning model, and iteratively improve your model.\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: "you will techniques"

File: learntools/machine_learning/nbs/tut2.ipynb (hunk 4 above)
Comment (colinmorris) [reply learntools/machine_learning/nbs/tut2.ipynb]: You might even defer talking about the Iowa data until they're actually in the Exercise notebook. (I definitely wouldn't get as specific as talking about details like the Iowa file path at this point).

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    "from learntools.machine_learning.ex2 import *\n",
+    "print(\"Setup Complete\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 1: Loading Data\n",
+    "Read the file into a Pandas DataFrame called `home_data`."
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: "the file" is a bit vague. Maybe "Read the Iowa housing data"? or "Read the Iowa data file"?

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    "# step_1.hint()\n",
+    "# step_1.solution()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 2: Review The Data\n",
+    "Use the command you learned to view summary statistics of the data. Then fill in variables to answer the following questions"
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: The example code in the tutorial does `print(df.describe())`. If they do that on the Iowa data, they get an abbreviated output with only 4 columns (which doesn't include the columns they need to answer this question). But if they just type `df.describe()`, they'll get a nice horizontally-scrolling table with all the columns. I'd suggest changing the tutorial example to not use print.

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# What is the average lot size (rounded to nearest integer)?\n",
+    "ave_lot_size = _\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: "ave"

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "#step_2.hint()\n",
+    "#step_2.solution()"
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: The solution code doesn't include `size_of_more_common_relation_space`.

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Not all 38 columns are shown in the output (indicated by the `...` in the middle of the output). There are too many columns to show everything.\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: This seems like a non-sequitur. Is this cell supposed to be here?

File: learntools/machine_learning/nbs/ch2.ipynb [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+   "source": [
+    "## Think About Your Data\n",
+    "\n",
+    "The newest house in your data isn't that new.  A few potential explanations for this:\n",
+    "1. They haven't built new houses where this data was collected.\n",
+    "1. The data was collected a long time ago. Houses built after the data publication wouldn't show up.\n",
+    "\n",
+    "If the reason is explanation #1 above, does that affect your trust in the model you build with this data? What about if it is reason #2?\n",
+    "\n",
+    "How you might check whether one explanation is more than the other?\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch2.ipynb]: more likely?

Comment (colinmorris) [reply learntools/machine_learning/nbs/ch2.ipynb]: Also "How you might" -> "How might you"?

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+   "metadata": {
+    "_cell_guid": "dc1e7b02-f022-48ca-a270-21f4b0c70835",
+    "_uuid": "0696f41450f17b2e0ebbbbb30a29053cde7de211"
+   },
+   "source": [
+    "**[Machine Learning Course Home Page](kaggle.com/learn/machine-learning).**\n",
+    "\n",
+    "---\n",
+    "\n",
+    "# Selecting and Data for Modeling\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: Title feels a bit awkward. At first glance, it seems like a typo for "Selecting Data for Modeling" (maybe it is?)

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "\n",
+    "melbourne_file_path = '../input/melbourne-housing-snapshot/melb_data.csv'\n",
+    "melbourne_data = pd.read_csv(melbourne_file_path) \n",
+    "print(melbourne_data.columns)"
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: This is sort of a matter of preference, but if I want to display one thing from one cell, I prefer to just have it as a naked expression at the end of the cell (i.e. just `melbourne_data.columns`) rather than explicitly printing it. Usually the results are either the same or better (because if you omit the print, you can get the benefit of objects that implement rich IPython repr methods).

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The Melbourne data has somemissing values (some houses for which some variables weren't recorded.)\n",
+    "# We'll learn to handle missing values in a later tutorial.  \n",
+    "# Your Iowa data doesn't have missing values in the predictors you use. \n",
+    "# So we will take the simplest option for now, and drop those houses from our data. \n",
+    "#Don't worry about this much for now, though the code is:\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: If you think this really isn't important, what about just hiding the code cell?

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "_cell_guid": "a52b0b8d-ea0a-435b-bde1-1472f63977ba",
+    "_uuid": "a0fac0889241e77f2e2deefbd63bc9f057d0c826"
+   },
+   "source": [
+    "# Choosing Predictors\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: Maybe worth having a brief aside on terminology here? I'm imagining there might be readers who have some exposure to alternative terms like "features" or "independent variables", and it might take a while for it to click that what we're talking about here is the same thing.

Comment (colinmorris) [reply learntools/machine_learning/nbs/tut3.ipynb]: Also, whether or not you go into synonyms, I think you need some explanation of what we actually mean by "predictors". i.e. these are the inputs to our model, the variables it will consider when predicting the target variable (price). Our first simple decision tree model had one predictor, which was number of bedrooms. etc.

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+    "Let's quickly review the data we'll be using to predict house prices using the `describe` method and the `head` method, which shows the top few rows."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(X.describe())\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: Same note on printing

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+    "print(X.head())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Visually checking your data with these commands is an important part of a data scientist's job.  You'll frequently find surprises in the dataset that deserve further inspection (e.g. to verify the data you have is correct)."
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: I feel like this parenthetical belongs with the first sentence. (Or could just be cut)

File: learntools/machine_learning/nbs/ch3.ipynb [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+    "# from _ import _\n",
+    "#specify the model. \n",
+    "#For model reproducibility, set a numeric value for random_state when specifying the model\n",
+    "iowa_model = _\n",
+    "\n",
+    "# Fit the model\n",
+    "_\n",
+    "\n",
+    "step_3.check()"
+   ]
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch3.ipynb]: I tried tackling this step incrementally, first trying to fix the import then running the code cell to see if it worked. The Incorrect message I got was really weird. It said something like `Incorrect: Expected type object of type DecisionTreeRegressor but got an object of type Pandas.IndexedSeries`. When I ran it again, it said the type was `str`. It took me a step to wrap my head around this - what's going on is that the Python interpreter stores the value of the last expression in the variable `_`, so while it seems like `_` is just a blank line saying "your answer goes here", it actually has some very wacky implications. 

I don't know if it's worth changing at this point, but it's at least something to be aware of in the future (and in case it comes up in any forum posts or comments).

File: learntools/machine_learning/nbs/tut3.ipynb [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+    "There are many ways to select a subset of your data. The [Pandas Course](https://www.kaggle.com/learn/pandas) covers these in more depth. But we can get our data set up for modeling with just two of these approaches.\n",
+    "\n",
+    "1. Dot notation, which we use to select the \"prediction target\"\n",
+    "2. Selecting with a column list, which we use to select the \"predictors\"\n",
+    "\n",
+    "## Selecting The Prediction Target \n",
+    "You can pull out a variable with **dot-notation**.  This single column is stored in a **Series**, which is broadly like a DataFrame with only a single column of data.  \n",
+    "\n",
+    "We'll use the dot notation to select the column we want to predict, which is called the **prediction target**. By convention, the prediction target is called **y**. So the code we need to save the house prices in the Melbourne data is"
+   ]
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut3.ipynb]: Big fan of bolding keywords like this. I'm an inveterate skimmer (and I bet lots of our readers are too)

File: learntools/machine_learning/nbs/ch2.ipynb (hunk 9 above)
Comment (dansbecker) [reply learntools/machine_learning/nbs/ch2.ipynb]: I'm not sure what this comment is saying.

File: learntools/machine_learning/nbs/tut3.ipynb (hunk 13 above)
Comment (dansbecker) [reply learntools/machine_learning/nbs/tut3.ipynb]: Yep. Typo

File: learntools/machine_learning/nbs/tut3.ipynb (hunk 15 above)
Comment (dansbecker) [reply learntools/machine_learning/nbs/tut3.ipynb]: Planning to hide it, and had forgotten to do that. 

Additionally I'm leaving this comment so that it doesn't seem like I'm doing sleight-of-hand to anyone who opens up the hidden cell.

File: learntools/machine_learning/nbs/tut4.ipynb [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+    "\n",
+    "Many people make a huge mistake when measuring predictive accuracy. They make predictions with their *training data* and compare those predictions to the target values in the *training data*. You'll see the problem with this approach and how to solve it in a moment, but let's think about how we'd do this first.\n",
+    "\n",
+    "You'd first need to summarize the model quality into an understandable way. If you compare predicted and actual home values for 10000 houses, you'll likely find mix of good and bad predictions. Looking through a list of 10000 predicted and actual values would be pointless. We need to summarize this into a single metric.\n",
+    "\n",
+    "There are many metrics for summarizing model quality, but we'll start with one called Mean Absolute Error (also called MAE). Let's break down this metric starting with the last word, error.\n",
+    "\n",
+    "The prediction error for each house is: <br>\n",
+    "error=actual−predicted\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut4.ipynb]: Maybe format in a code block? Or as latex?

File: learntools/machine_learning/nbs/tut4.ipynb [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+    " \n",
+    "So, if a house cost \\$150,000 and you predicted it would cost \\$100,000 the error is \\$50,000.\n",
+    "\n",
+    "With the MAE metric, we take the absolute value of each error. This converts each error to a positive number. We then take the average of those absolute errors. This is our measure of model quality. In plain English, it can be said as\n",
+    "\n",
+    "On average, our predictions are off by about X\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut4.ipynb]: Blockquote?

File: learntools/machine_learning/nbs/tut4.ipynb [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+    "# run this script.\n",
+    "train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)\n",
+    "# Define model\n",
+    "melbourne_model = DecisionTreeRegressor()\n",
+    "# Fit model\n",
+    "melbourne_model.fit(train_X, train_y)\n",
+    "\n",
+    "# get predicted prices on validation data\n",
+    "val_predictions = melbourne_model.predict(val_X)\n",
+    "print(mean_absolute_error(val_y, val_predictions))"
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut4.ipynb]: We kind of breeze past the result of this experiment, but I think it's really worth remarking on. Going from being off by $435 on average to being off by $263,519 is an incredible difference, and a dramatic demonstration of why validating on training data is so dangerous.

It's such a huge difference, it makes me really curious to learn more. How did our model overfit our data so precisely? What does our decision tree actually look like? Presumably it's really deep? How do the number of leaves compare to the number of training instances? What features is it overfitting to? (i.e. what's the analogy to the "green door" in our case?) Maybe this would be worth investigating a bit in the next tutorial (underfitting and overfitting)?

File: learntools/machine_learning/nbs/ch4.ipynb [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+      ],
+      "text/plain": [
+       "Incorrect: You still need to define the following variables: `train_X`, `val_y`"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch4.ipynb]: The notebook on the site has a stray `_` just hanging out on the second line of this cell.

Comment (colinmorris) [reply learntools/machine_learning/nbs/ch4.ipynb]: It also has a typo (traaining)

File: learntools/machine_learning/nbs/ch4.ipynb [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+    "# step_1.solution()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 2: Specify and Fit the Model\n",
+    "\n",
+    "Create a `DecisionTreeRegressor` model and fit it to the relevant data.\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch4.ipynb]: I would have said "to the training data". Or are you being intentionally vague to force them to think about the fact that we're now separating the training and validation data?

File: learntools/machine_learning/nbs/ch4.ipynb [hunk 26]
```diff
@@ -0,0 +1,310 @@
...
+   ],
+   "source": [
+    "# You imported DecisionTreeRegressor in your last exercise\n",
+    "# and that code has been copied to the setup code above. So, no need to\n",
+    "# import it again\n",
+    "\n",
+    "# Specify the model\n",
+    "iowa_model = _\n",
+    "\n",
+    "# Fit iowa_model with the training data.\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch4.ipynb]: FYI, they still get marked as correct if they train on the full data. Maybe this is hard to check for though.

File: learntools/machine_learning/nbs/ch2.ipynb (hunk 9 above)
Comment (colinmorris) [reply learntools/machine_learning/nbs/ch2.ipynb]: Sorry just pointing out what I think is a typo (should be `avg_lot_size`, right?)

File: learntools/machine_learning/nbs/tut4.ipynb (hunk 23 above)
Comment (colinmorris) [reply learntools/machine_learning/nbs/tut4.ipynb]: Also (sorry, I can't stop thinking about this), an average error of $263k seems *bad*. Is it? What MAE do we get by just guessing the overall training-set average price every time? Is this heavily influenced by outliers?

I like that the exercise after this has us look at a few examples of predictions on the validation data and compare them to the actual price. Could be interesting to do that here to get a feel for the kinds of errors we're dealing with.

**Comment by dansbecker** [comment]
I've always used ave for average.  Is it your sense that avg is much more
common for something like this?  If our bq connection to the github data
were more reliable, it'd be interesting to run some sort of comparison
(e.g. number of occurances between `ave_price` vs `avg_price`)

File: learntools/machine_learning/nbs/ch4.ipynb (hunk 26 above)
Comment (colinmorris) [reply learntools/machine_learning/nbs/ch4.ipynb]: That in itself is maybe not a huge deal. But if they do make this mistake, they'll be able to continue on until step 4, then it will finally catch up with them because we'll say their `val_mae` value is incorrect. If someone does get to that point, it might be hard for them to figure out that they need to retrace their steps to step 2.

File: learntools/machine_learning/nbs/ch2.ipynb (hunk 9 above)
Comment (colinmorris) [reply learntools/machine_learning/nbs/ch2.ipynb]: (Replying to your response here, since it looks like Github put it as a top-level comment that I can't reply to)

Wow, that's really surprising to me. I didn't even consider the possibility that it wasn't a typo. I don't think I've ever seen "ave" as an abbreviation for "average". Maybe it's another econometrics thing :)

File: learntools/machine_learning/nbs/tut5.ipynb [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+    "_cell_guid": "91c024da-525f-496d-90a5-ac395dd5b87b",
+    "_uuid": "6751d138eff1d9eca94146995ce84554132177d5",
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# compare MAE with differing values of max_leaf_nodes\n",
+    "for max_leaf_nodes in [5, 50, 500, 5000]:\n",
+    "    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)\n",
+    "    print(\"Max leaf nodes: %d  \\t\\t Mean Absolute Error:  %d\" %(max_leaf_nodes, my_mae))"
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut5.ipynb]: In the kernel this cell is giving a NameError currently (on `train_X`).

File: learntools/machine_learning/nbs/tut5.ipynb [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+    "\n",
+    "---\n",
+    "\n",
+    "At the end of this step, you will understand the concepts of underfitting and overfitting, and you will be able to apply these ideas to make your models more accurate.\n",
+    "\n",
+    "# Experimenting With Different Models\n",
+    "\n",
+    "Now that you have a reliable way to measure model accuracy, you can experiment with alternative models and see which gives the best predictions.  But what alternatives do you have for models?\n",
+    "\n",
+    "You can see in scikit-learn's [documentation](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html) that the decision tree model has many options (more than you'll want or need for a long time). The most important options determine the tree's depth.  Recall from [page 2](https://www.kaggle.com/dansbecker/first-data-science-scenario-page-2/) that a tree's depth is a measure of how many splits it makes before coming to a prediction.  This is a relatively shallow tree\n",
```
[the remark this answers is not in the document]
Comment (dansbecker) [reply learntools/machine_learning/nbs/tut5.ipynb]: Woah.  That link is from 2017.  Before we'd launched kaggle.com/Learn.  I'll figure out where it links to and fix it, but I'm a little troubled that it's even there still.

File: learntools/machine_learning/nbs/tut5.ipynb [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "_cell_guid": "5f2eb504-a4e5-4857-a974-6830ff34c042",
+    "_uuid": "98c63957ec7c7c7f235e659ca5e6a2aea6e36ad3"
+   },
+   "source": [
+    "\n",
+    "Of the options listed, 500 is the optimal number of leaves.  Apply the function to your Iowa data to find the best decision tree.\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut5.ipynb]: Last sentence is kind of funny. Should it be something like "In the next exercise, you will..."?

File: learntools/machine_learning/nbs/tut5.ipynb [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+    "\n",
+    "# Conclusion\n",
+    "\n",
+    "Here's the takeaway: Models can suffer from either:\n",
+    "- **Overfitting:** capturing spurious patterns that won't recur in the future, leading to less accurate predictions, or \n",
+    "- **Underfitting:** failing to capture relevant patterns, again leading to less accurate predictions. \n",
+    "\n",
+    "We use **validation** data, which isn't used in model training, to measure a candidate model's accuracy. This lets us try many candidate models and keep the best one. \n",
+    "\n",
+    "But we're still using Decision Tree models, which are not very sophisticated by modern machine learning standards. \n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut5.ipynb]: Was this intended as a segue into the next lesson? If so, might be worth being more explicit about that

File: learntools/machine_learning/nbs/ch6.ipynb [hunk 31]
```diff
@@ -0,0 +1,172 @@
...
+   "source": [
+    "# Exercises\n",
+    "Data science isn't always this easy. But replacing the decision tree with a Random Forest is going to be an easy win."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 1: Use a Random Forest"
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch6.ipynb]: Having this heading feels a bit silly given that there's only one step.

Overall this exercise does feel quite short compared to the previous ones. I don't think it's a blocker, but it might be worth thinking about adding another step or two later (like suggesting a couple `RandomForestRegressor` params for them to experiment with).

File: learntools/machine_learning/nbs/tut7.ipynb [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "from sklearn.ensemble import RandomForestRegressor\n",
+    "\n",
+    "# Read the data\n",
+    "train = pd.read_csv('../input/train.csv')\n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut7.ipynb]: This is giving file not found errors on the site.

File: learntools/machine_learning/nbs/tut7.ipynb [hunk 33]
```diff
@@ -0,0 +1,154 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "_cell_guid": "ab2280a4-9c65-4810-8813-831f4dcd86e1",
+    "_uuid": "97f80b7f76066fd78cdb5137757fb3c63abd0519"
+   },
+   "source": [
+    "# Make Submission\n",
+    "Hit the blue **Commit and Run** button at the top of your notebook screen.  It will take some time for your kernel to run.  When it has finished your navigation bar at the top of the screen will have a tab for **Output**.  This only shows up if you have written an output file (like we did in the **Prepare Submission File** step).  \n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut7.ipynb]: Reading these instructions makes me wonder: was I supposed to fork this notebook at the beginning?

File: learntools/machine_learning/nbs/tut7.ipynb [hunk 34]
```diff
@@ -0,0 +1,154 @@
+{
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/tut7.ipynb]: I'm finding this last lesson hard to follow. Thoughts I had during my first readthrough:

- We're going to go through an example of submitting to a competition. Okay, *which competition*?
- Are we still working with the Melbourne housing data? Seems like the predictor names are the same as before, but the filenames are different (`train.csv` and `test.csv` instead of `melb_data.csv`)

The first 6 lessons did a great job of building on top of each other and following a coherent thread, but this one feels disconnected. 

I think part of the confusion is that the first 6 lessons were about explaining machine learning concepts, but this one is explaining the interface around a specific Kaggle feature. It almost seems like this is a piece of Kaggle documentation.

File: learntools/machine_learning/nbs/ch7.ipynb [hunk 35]
```diff
@@ -0,0 +1,160 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# To improve accuracy, create new Random Forest model which you will train on all training data\n",
+    "rf_model_on_full_data = _\n",
+    "\n",
+    "# fit rf_model_on_full_data on all data from the \n",
```
Comment (colinmorris) [inline learntools/machine_learning/nbs/ch7.ipynb]: from the... competition training data?

File: learntools/machine_learning/nbs/tut7.ipynb (hunk 34 above)
Comment (colinmorris) [reply learntools/machine_learning/nbs/tut7.ipynb]: That said, conceptually, I do like the idea of using a competition as the carrot to induce them to experiment and try to push their models further. I just wonder if this notebook should focus more on the ML stuff, with the tutorial on the competitions interface going in a separate standalone notebook (linked to from the tutorial notebook).

One example of a more ML-ish thing we could talk about here: When we submit our predictions to the competition, we see that we earned a score like 0.19734, which looks nothing like the MAEs that we were getting. Digging in, we see that this competition is ranked by Root Mean Squared Logarithmic Error. What's that? How can we measure it on our validation set? Should we re-assess our choice of parameters if we're optimizing for this new metric?

File: learntools/machine_learning/nbs/tut1.ipynb (hunk 2 above)
Comment (dansbecker) [reply learntools/machine_learning/nbs/tut1.ipynb]: Indeed.  This (and most issues you identified) were in the live version online too. Going back through everything, I'm amazed how much worse these materials were than I'd thought.
